### PR TITLE
Exhaustively testing SimpleRandomSample, software eng improvements

### LINKED ISSUE #75: Negative float as sampsize and popsize
There is another issue in SurveyDesign.jl. If you write the following code, then the error message showing is not what it should be
`srs = SimpleRandomSample(apisrs, popsize = -2.8, ignorefpc = true)`
the error message is 

> either population size or frequency/probability weights must be specified

Also for this code 
`    srs = SimpleRandomSample(apisrs, sampsize = -2.8, ignorefpc = true)# the function is working up to line 55`
the function is working all the way up to the 55th line and then only shows that there is an error giving the following message:

> ERROR: InexactError: UInt64(-3.0)

My suggestion is we add some check that lets us proceed with the function only if the sampsize and popsize are non-negative integers.

### PR DESCRIPTION
**Pull request by smishr**
I have been making lots of software engineering improvements and if-else ladder type changes as we move towards 0.2 package release. This PR also attempts to continue and resolve #75 #67 #70 #64.

Specifically, `SimpleRandomSample` is being exhaustively tested and reformatted, with more type checking and intelligent parsing behaviour.

### CODE REVIEW

**Comment by smishr** [comment]
@ayushpatnaikgit After the latest commits, it passes `]test Survey` locally on my Mac

File: src/SurveyDesign.jl [hunk 1]
```diff
@@ -15,36 +15,102 @@ abstract type AbstractSurveyDesign end
...
         ignorefpc=false
     )
+        # Only valid argument types given to constructor
+        argtypes_weights = Union{Nothing,Symbol,Vector{<:Real}}
+        argtypes_probs = Union{Nothing,Symbol,Vector{<:Real}}
+        argtypes_popsize = Union{Nothing,Symbol,<:Unsigned,Vector{<:Real}}
+        argtypes_sampsize = Union{Nothing,Symbol,<:Unsigned,Vector{<:Real}}
+        # If any invalid type raise error
+        if !(isa(weights, argtypes_weights))
+            error("Invalid type of argument given for `weights` argument")
```
Comment (ayushpatnaikgit) [inline src/SurveyDesign.jl:58]:
```suggestion
            error("invalid type of argument given for `weights` argument")
```

File: src/SurveyDesign.jl [hunk 2]
```diff
@@ -15,36 +15,102 @@ abstract type AbstractSurveyDesign end
...
+        elseif !(isa(probs, argtypes_probs))
+            error("Invalid type of argument given for `probs` argument")
```
Comment (ayushpatnaikgit) [inline src/SurveyDesign.jl:60]:
```suggestion
            error("invalid type of argument given for `probs` argument")
```

File: src/SurveyDesign.jl [hunk 3]
```diff
@@ -15,36 +15,102 @@ abstract type AbstractSurveyDesign end
...
+        elseif !(isa(popsize, argtypes_popsize))
+            error("Invalid type of argument given for `popsize` argument")
```
Comment (ayushpatnaikgit) [inline src/SurveyDesign.jl:62]:
```suggestion
            error("invalid type of argument given for `popsize` argument")
```

File: src/SurveyDesign.jl [hunk 4]
```diff
@@ -15,36 +15,102 @@ abstract type AbstractSurveyDesign end
...
+        elseif !(isa(sampsize, argtypes_sampsize))
+            error("Invalid type of argument given for `sampsize` argument")
```
Comment (ayushpatnaikgit) [inline src/SurveyDesign.jl:64]:
```suggestion
            error("invalid type of argument given for `sampsize` argument")
```

File: src/SurveyDesign.jl [hunk 5]
```diff
@@ -53,23 +119,46 @@ struct SimpleRandomSample <: AbstractSurveyDesign
...
-            if !all(y -> y == first(popsize), popsize) # SRS by definition is equi-weighted
-                error("Simple Random Sample must be equi-weighted. Different sampling weights detected in vectors")
-            end
-            weights = popsize ./ sampsize # ratio estimator for SRS
-            popsize = first(popsize) |> UInt
+        elseif typeof(popsize) <: Unsigned
+            weights = fill(popsize / sampsize, nrow(data)) # If popsize is given, weights vector is made concordant with popsize and sampsize, regardless of given weights argument
         else
-            error("either population size or frequency/probability weights must be specified")
+            error("Something went wrong. Please check validity of inputs.")
```
Comment (ayushpatnaikgit) [inline src/SurveyDesign.jl:134]:
```suggestion
            error("something went wrong, please check validity of inputs.")
```

File: src/SurveyDesign.jl [hunk 6]
```diff
@@ -53,23 +119,46 @@ struct SimpleRandomSample <: AbstractSurveyDesign
...
+        # TODO: This works under some cases, but should find better way to process ignoring fpc
+        if ignorefpc
+            @warn "assuming all weights are equal to 1.0"
+            weights = ones(nrow(data))
+            probs = 1 ./ weights
         end
+        # sum of weights must equal to `popsize` for SRS
+        if !isnothing(weights) && !(isapprox(sum(weights), popsize; atol=1e-4))
+            if ignorefpc && !(isapprox(sum(weights), sampsize; atol=1e-4)) # Change if ignorefpc functionality changes
+                error("Sum of sampling weights should be equal to `sampsize` for Simple Random Sample with ignorefpc")
```
Comment (ayushpatnaikgit) [inline src/SurveyDesign.jl:146]:
```suggestion
                error("sum of sampling weights should be equal to `sampsize` for `SimpleRandomSample` with `ignorefpc`")
```

File: src/SurveyDesign.jl [hunk 7]
```diff
@@ -53,23 +119,46 @@ struct SimpleRandomSample <: AbstractSurveyDesign
...
+            elseif !ignorefpc
+                @show sum(weights)
+                error("Sum of sampling weights must be equal to `popsize` for Simple Random Sample")
```
Comment (ayushpatnaikgit) [inline src/SurveyDesign.jl:149]:
```suggestion
                error("sum of sampling weights must be equal to `popsize` for `SimpleRandomSample`")
```

File: src/SurveyDesign.jl [hunk 8]
```diff
@@ -53,23 +119,46 @@ struct SimpleRandomSample <: AbstractSurveyDesign
...
+            end
+        end
+        # sum of probs must equal popsize for SRS
+        if !isnothing(probs) && !(isapprox(sum(1 ./ probs), popsize; atol=1e-4))
+            if ignorefpc && !(isapprox(sum(1 ./ probs), sampsize; atol=1e-4)) # Change if ignorefpc functionality changes
+                error("Sum of inverse sampling probabilities should be equal to `sampsize` for Simple Random Sample with ignorefpc")
```
Comment (ayushpatnaikgit) [inline src/SurveyDesign.jl:155]:
```suggestion
                error("sum of inverse sampling probabilities should be equal to `sampsize` for `SimpleRandomSample` with `ignorefpc`")
```

File: src/svymean.jl [hunk 9]
```diff
@@ -54,7 +54,7 @@ function svymean(x::Symbol, design::SimpleRandomSample)
         p.se = sqrt.(p.var)
         return p
     end
-    return DataFrame(mean = mean(design.data[!, x]), sem = sem(x, design::SimpleRandomSample))
+    return DataFrame(mean=mean(design.data[!, x]), sem=sem(x, design::SimpleRandomSample))
```
Comment (ayushpatnaikgit) [inline src/svymean.jl:57]: Is this the correct style?

**Review by ayushpatnaikgit — COMMENTED** [review]
Changed error messages to begin all of them with small letters.

File: src/svymean.jl (hunk 9 above)
Comment (smishr) [reply src/svymean.jl:57]: i ran julia formatter so that should make these lines  consistent

**Pull request merged**